Mathlib/GroupTheory/Index.lean

```diff
@@ -213,6 +213,63 @@ theorem index_eq_two_iff : H.index = 2 ↔ ∃ a, ∀ b, Xor' (b * a ∈ H) (b 
     exact one_mem _
   · rwa [ha, inv_mem_iff (x := b)]
 
+/-- A subgroup has index two if and only if there exists `a` such that for all `b`, exactly one
+of `a * b` and `b` belong to `H`. -/
+@[to_additive /-- An additive subgroup has index two if and only if there exists `a` such that
+for all `b`, exactly one of `a + b` and `b` belong to `H`. -/]
+theorem index_eq_two_iff' : H.index = 2 ↔ ∃ a, ∀ b, Xor' (a * b ∈ H) (b ∈ H) := by
+  rw [index_eq_two_iff, (Equiv.inv G).exists_congr]
+  refine fun a ↦ (Equiv.inv G).forall_congr fun b ↦ ?_
+  simp only [Equiv.inv_apply, inv_mem_iff, ← mul_inv_rev]
+
+/-- A subgroup `H` has index two if and only if there exists `a ∉ H` such that for all `b`, one
+of `b * a` and `b` belongs to `H`. -/
+@[to_additive /-- An additive subgroup `H` has index two if and only if there exists `a ∉ H` such
+that for all `b`, one of `b + a` and `b` belongs to `H`. -/]
+lemma index_eq_two_iff_exists_notMem_and :
+    H.index = 2 ↔ ∃ a, a ∉ H ∧ ∀ b, (b * a ∈ H) ∨ (b ∈ H) := by
```

Comment:
To me this feels a bit strange. The `a ∉ H` makes the RHS stronger, but the `∨` rather than the `Xor'` makes the RHS weaker. In my mind, it would really only make sense to have two versions, one where the RHS is as weak as possible (e.g., just `∃ a, ∀ b, (b * a ∈ H) ∨ (b ∈ H)`) and one where the RHS is as strong as possible (e.g., `∃ a, a ∉ H ∧ ∀ b, Xor' (a * b ∈ H) (b ∈ H)`). And likewise for the relindex versions.

Reply:
The "weak as possible" formulation you propose, `∃ a, ∀ b, (b * a ∈ H) ∨ (b ∈ H)`, is too weak: it is not equivalent to `H.index = 2` – it is also satisfied for `H.index = 1` (i.e. `H`=everything) as well. 

On the other hand, `∃ a, a ∉ H ∧ ∀ b, Xor' (a * b ∈ H) (b ∈ H)` is too strong – the first clause of the `and` is redundant, since it is implied by the second (specializing `∀ b, Xor' (a * b ∈ H) (b ∈ H)` to `b = 1` implies that `a ∉ H`). It would seem odd to me to put `P ∧ Q` in the statement of a lemma when `Q` implies `P`.

Reply:
(It would be nice if we could have 
```index_le_two_iff : H.index ≤ 2 ↔ ∃ a, ∀ b, (b * a ∈ H) ∨ (b ∈ H)```
but it doesn't work, because `Subgroup.index` is zero for infinite-index subgroups – a design decision I strongly dislike, precisely because it makes it impossible to reason about inequalities between subgroup indices in a natural way. But that is not a matter for this PR.)

Reply:
Yes, definitely, thanks!